scrapy/downloadermiddlewares/retry.py

```diff
@@ -23,12 +24,106 @@
 )
 from twisted.web.client import ResponseFailed
 
-from scrapy.exceptions import NotConfigured
-from scrapy.utils.response import response_status_message
 from scrapy.core.downloader.handlers.http11 import TunnelError
+from scrapy.exceptions import NotConfigured
+from scrapy.http.request import Request
+from scrapy.spiders import Spider
 from scrapy.utils.python import global_object_name
+from scrapy.utils.response import response_status_message
+
 
-logger = logging.getLogger(__name__)
+retry_logger = getLogger(__name__)
+
+
+def get_retry_request(
+    request: Request,
+    *,
+    spider: Spider,
+    reason: Union[str, Exception] = 'unspecified',
+    max_retry_times: Optional[int] = None,
+    priority_adjust: Union[int, float, None] = None,
+    logger: Logger = retry_logger,
+    stats_base_key: str = 'retry',
+):
+    """
+    Returns a new :class:`~scrapy.Request` object to retry the specified
+    request, or ``None`` if retries of the specified request have been
+    exhausted.
+
+    For example, in a :class:`~scrapy.Spider` callback, you could use it as
+    follows::
+
+        def parse(self, response):
+            if not response.text:
+                new_request = get_retry_request(
+                    response.request,
+                    spider=self,
+                    reason='empty',
+                )
+                if new_request:
```

Comment:
Considering that we can `yield None`, maybe you don't need the `if`

Reply:
You definitely could write it without the `if`. My main concern here is readability. What about doing what you suggest and renaming the variable as `new_request_or_none`?